crates/task-manager/src/response_calculator.rs

```diff
@@ -0,0 +1,120 @@
+use rand::Rng;
+use std::{future::Future, marker::PhantomData};
+use tracing::info;
+
+use crate::TaskManagerError;
+
+/// Trait used to compute the response of a task.
+pub trait ResponseCalculator<Input, Output> {
+    /// Compute the response of a task.
+    ///
+    /// # Arguments
+    ///
+    /// * `task_index` - The index of the task.
+    /// * `input` - The input of the task.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<Output, TaskManagerError>` - The response of the task.
+    fn compute_response(
+        &self,
+        task_index: u32,
+        input: Input,
+    ) -> impl Future<Output = Result<Output, TaskManagerError>>;
+}
+
+/// Implementation of the [`ResponseCalculator`] trait that uses a function to compute the response.
+#[derive(Debug)]
+pub struct FunctionResponseCalculator<Input, Output, F>
+where
+    F: AsyncFn(u32, Input) -> Result<Output, TaskManagerError>,
+{
+    /// The function used to compute the response.
+    pub compute_fn: F,
+    _input: PhantomData<Input>,
+    _output: PhantomData<Output>,
+}
```

Comment:
Can we simplify this to something like this?

```suggestion
#[derive(Debug)]
pub struct FunctionResponseCalculator<F>(F);
```

And move the bounds to the `impl`